### PR TITLE
nix: always depend on nix

### PR DESCRIPTION
**Pull request by malt3**
### Context

This makes nix a hard dependency for building Constellation.
This simplifies a lot of depdendency / toolchain handling.

### Proposed change(s)

- nix: always depend on nix
- use openssl from nixpkggs
- install nix in GitHub Actions workflows

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: dev-docs/workflows/build-develop-deploy.md [hunk 1]
```diff
@@ -7,35 +7,21 @@ Prerequisites:
 * 20GB (minimum), better 40 GB disk space (required if you want to cross compile for all platforms)
 * [Latest version of Go](https://go.dev/doc/install).
 * Unless you use Nix / NixOS: [Bazelisk installed as `bazel` in your path](https://github.com/bazelbuild/bazelisk/releases).
-* We recommend Nix installed via [determinate systems installer](https://github.com/DeterminateSystems/nix-installer) (or NixOS as host system).
+* We require Nix installed via [determinate systems installer](https://github.com/DeterminateSystems/nix-installer) (or NixOS as host system).
```
Comment (katexochen) [inline dev-docs/workflows/build-develop-deploy.md:10]: Hm, we required Nix installed, but should work with any installation, shouldn't it? So maybe separate requirement and recommendation here.

File: .github/actions/setup_bazel_nix/action.yml [hunk 2]
```diff
@@ -25,6 +26,55 @@ runs:
...
+
+    - name: Install Bazelisk
+      shell: bash
+      if: steps.check_inputs.outputs.bazelPreinstalled == 'false' && steps.check_inputs.outputs.nixOS == 'false'
+      env:
+        OS: ${{ steps.check_inputs.outputs.os }}
+        ARCH: ${{ steps.check_inputs.outputs.arch }}
+      run: |
+        echo "::group::Install Bazelisk"
+        curl -sSfLo /usr/local/bin/bazel "https://github.com/bazelbuild/bazelisk/releases/download/v1.18.0/bazelisk-${OS}-${ARCH}"
```
Comment (katexochen) [inline .github/actions/setup_bazel_nix/action.yml:76]:
```suggestion
        curl -fsSLo /usr/local/bin/bazel "https://github.com/bazelbuild/bazelisk/releases/download/v1.18.0/bazelisk-${OS}-${ARCH}"
```
Just for consistency.

File: .github/actions/setup_bazel_nix/action.yml [hunk 3]
```diff
@@ -25,6 +26,55 @@ runs:
...
+        fi
+        echo "::endgroup::"
+
+    - name: Install nix
+      uses: cachix/install-nix-action@6ed004b9ccb68dbc28e7c85bee15fa93dbd214ac # v22
+      if: steps.check_inputs.outputs.nixPreinstalled == 'false'
+
+    - name: Install Bazelisk
+      shell: bash
+      if: steps.check_inputs.outputs.bazelPreinstalled == 'false' && steps.check_inputs.outputs.nixOS == 'false'
```
Comment (katexochen) [inline .github/actions/setup_bazel_nix/action.yml:70]: nit: For this step and the one above, try to keep a conventional ordering:

name
id
if

uses/shell
env
...
runs

File: .github/actions/setup_bazel_nix/action.yml [hunk 4]
```diff
@@ -1,5 +1,5 @@
-name: Setup bazel
-description: Setup Bazel for CI builds and tests
+name: Setup bazel and Nix
```
Comment (katexochen) [inline .github/actions/setup_bazel_nix/action.yml:1]: Can we keep the cachix step and comment it out? I guess we will need it from time to time.

**Review by katexochen — COMMENTED** [review]
Some nits upfront, will review the rest tomorrow.

File: WORKSPACE.bazel [hunk 5]
```diff
@@ -28,6 +28,28 @@ nixpkgs_flake_package(
...
+    name = "org_openssl",
+    srcs = glob(["lib/**/*.a"]),
+    hdrs = [":include"],
+    strip_include_prefix = "include",
+    visibility = ["//visibility:public"],
+)
+""",
+    nix_flake_file = "//:flake.nix",
+    nix_flake_lock_file = "//:flake.lock",
+    package = "openssl",
```
Comment (katexochen) [inline WORKSPACE.bazel:50]: openssl has a .dev output. Do we need the full output?

File: flake.nix [hunk 6]
```diff
@@ -40,10 +35,17 @@
         ]);
       }));
 
+      openssl-static = pkgsUnstable.openssl.override { static = true; };
+
     in
     {
       packages.mkosi = mkosiDev;
 
+      packages.openssl = pkgsUnstable.symlinkJoin {
```
Comment (katexochen) [inline flake.nix:44]: What is happening here?

File: WORKSPACE.bazel (hunk 5 above)
Comment (malt3) [reply WORKSPACE.bazel:50]: This is using the `openssl` from flake.nix that combines multiple outputs. The `.dev` output only contains headers. We need the combined `.dev` (headers) and `.out` (compiled library) in one derivation for Bazel to be happy.

File: flake.nix (hunk 6 above)
Comment (malt3) [reply flake.nix:44]: `symlinkJoin` is an efficient method to combine the output of two derivations. This way, we get the headers and library (`.a`) in one derivation that can be used by `rules_nixpkgs` / Bazel cc_library rules.

This idea is also used here: https://github.com/tweag/rules_nixpkgs/blob/7e627d76ba65d6c42586fc265e46bd370672b4eb/examples/toolchains/rust/openssl-static.nix

**Review by katexochen — APPROVED** [review]

**Comment by malt3** [comment]
Currently waiting for [this commit](https://github.com/NixOS/nixpkgs/commit/187f681cb4921d413ad940834baab18654cb0f9a) to trickle into unstable. Otherwise, builds on macOS will fail.

**Pull request merged**